include/micm/solver/backward_euler.inl

```diff
@@ -0,0 +1,205 @@
+        // (y_{n+1} - y_n) / H = f(t_{n+1}, y_{n+1})
+
+        // try to find the root by factoring and solving the linear system
+        linear_solver.Factor(state.jacobian_, state.lower_matrix_, state.upper_matrix_, singular);
+
+        auto yn1_iter = Yn1.begin();
+        auto yn_iter = Yn.begin();
+        auto residual_iter = residual.begin();
+        auto forcing_iter = forcing.begin();
+        // forcing_blk in camchem
+        // residual = (Yn1 - Yn) / H - forcing;
+        for(; yn1_iter != Yn1.end(); ++yn1_iter, ++yn_iter, ++residual_iter, ++forcing_iter) {
+          *residual_iter = (*yn1_iter - *yn_iter) / H - *forcing_iter;
+        }
+
+        // the result of the linear solver will be stored in temp
+        // this represnts the change in the solution
+        linear_solver.Solve(residual, temp, state.lower_matrix_, state.upper_matrix_);
+
+        // solution_blk in camchem
+        // Yn1 = Yn1 + temp;
+        auto temp_iter = temp.begin();
+        yn1_iter = Yn1.begin();
+
+        for(; temp_iter != temp.end(); ++temp_iter, ++yn1_iter) {
+          *yn1_iter += *temp_iter;
+        }
+
+        // always make sure the solution is positive regardless of which iteration we are on
+        // remove negatives, accept this solution and continue
+        yn1_iter = Yn1.begin();
+        yn_iter = Yn.begin();
+        for(; yn1_iter != Yn1.end(); ++yn1_iter, ++yn_iter) {
+          *yn_iter = *yn1_iter = std::max(0.0, *yn1_iter);
+        }
+
+        if (iterations == 0) {
+          // we always want to accept the first iteration
+          continue;
+        }
+
+        // check for convergence
+        temp_iter = temp.begin();
+        yn1_iter = Yn1.begin();
+
+        converged = true;
+        // convergence happens when the absolute value of the change to the solution
+        // is less than a tolerance times the absolute value of the solution
+        for(; temp_iter != temp.end(); ++temp_iter, ++yn1_iter) {
+          // changes that are much smaller than the tolerance are negligible and we assume can be accepted
+          if (std::abs(*temp_iter) > small) {
+            if (std::abs(*temp_iter) > tol * std::abs(*yn1_iter)) {
+              converged = false;
+              break;
+            }
+          }
+        }
+
+        ++iterations;
+        // correct this check
+      } while(!converged && iterations < max_iter);
+
+      if (!converged) {
+        n_successful_integrations = 0;
+
+        if (n_convergence_failures >= time_step_reductions.size()) {
+          // we have failed to converge, accept the solution
+          // TODO: continue on with the current solution to get the full solution
+          n_convergence_failures = 0;
+          // give_up = true;
+          t += H;
+          break;
+          throw std::system_error(make_error_code(MicmBackwardEulerErrc::FailedToConverge), "Failed to converge");
+        };
+
+        H *= time_step_reductions[n_convergence_failures++];
```

Comment:
After doing the time substepping here, we need to reset `Yn` as:
```
          Yn = yn_bk;
```
since `Yn` has already been updated during this iteration.

Reply:
Is this the same thing as reseting Yn to the initial value for this timestep?

Reply:
I think so.